### ISSUE TITLE
Unable to use Flexdash on FF hosted Node-RED

### ISSUE THREAD

**Issue by knolleary** [FlowFuse/flowfuse#1778]
### Current Behavior

When attempting to access a Flexdash dashboard hosted by Node-RED within FF, the following error is hit causing Node-RED to restart:
```

TypeError: req.session.touch is not a function
    at ServerResponse.end (/Users/nol/code/ff/flowforge-dev-env/packages/flowforge/var/stacks/3.0.0/node_modules/express-session/index.js:330:21)
    at ServerResponse.end (/Users/nol/code/ff/flowforge-dev-env/packages/flowforge/var/projects/1e7bb9d0-36e5-40dd-888f-6927a635bc5b/node_modules/compression/index.js:107:21)
    at write (/Users/nol/code/ff/flowforge-dev-env/packages/flowforge/var/stacks/3.0.0/node_modules/finalhandler/index.js:302:9)
    at AsyncResource.runInAsyncScope (node:async_hooks:203:9)
    at listener (/Users/nol/code/ff/flowforge-dev-env/packages/flowforge/var/stacks/3.0.0/node_modules/on-finished/index.js:170:15)
    at onFinish (/Users/nol/code/ff/flowforge-dev-env/packages/flowforge/var/stacks/3.0.0/node_modules/on-finished/index.js:101:5)
    at callback (/Users/nol/code/ff/flowforge-dev-env/packages/flowforge/var/stacks/3.0.0/node_modules/ee-first/index.js:55:10)
    at IncomingMessage.onevent (/Users/nol/code/ff/flowforge-dev-env/packages/flowforge/var/stacks/3.0.0/node_modules/ee-first/index.js:93:5)
    at IncomingMessage.emit (node:events:525:35)
    at endReadableNT (node:internal/streams/readable:1358:12)
```

This is related to the Express Session handling within Node-RED - which our authentication plugin drives in someway to cause this incompatibility.

It isn't immediately clear if the required fix will sit in Node-RED, our auth plugin, or flexdash. I am working on the assumption it will be one of the first two - I don't see flexdash doing anything unreasonable here.

### Expected Behavior

Flexdash works

### Steps To Reproduce

- Install Flexdash within a FF hosted Node-RED
- Add the example flow flexdash provides
- Deploy and attempt to access the dashboard

**Comment by Steve-Mcl** [comment]
@tve, do you have any ideas what might be causing the error "req.session.touch is not a function"

Stack overflow send to suggest it is something to do with an express parser: https://stackoverflow.com/a/71030953

Note: this is happening consistently on the FlowForge platform where we have SSO enabled so as nick suggests it lies somewhere between 3 components. 

From what I can see from a cursory glance at flexdash source, I can't see where you are `use`ing any additional parsers or even touching express!

**Comment by knolleary** [comment]
@Steve-Mcl this is for us/Node-RED to resolve. There's a lot going on with the express sessions to get our SSO working. We don't use full session objects and attach just enough of an object for our SSO needs. For whatever reason, flexdash manages to trigger some interaction with the session object that isn't supported by the cutdown session we have.

The solution will be to use a more fully formed session object in our code. The difficulty is that some of that is on the Node-RED codebase as well as ours. The primary task is to pick that apart so we can apply the right fix to the right places.

**Comment by tve** [comment]
In case you haven't already spotted it, the main Express and Socket.io munging code is at https://github.com/flexdash/node-red-flexdash/blob/main/flexdash-dashboard.js#L231-L339

In order to run the dev server (vite) there is also some proxying done at https://github.com/flexdash/node-red-flexdash/blob/main/flexdash-dev-server.js#L200-L249 and I don't know whether this can work at all in FF.

**Comment by Steve-Mcl** [comment]
> In case you haven't already spotted it, the main Express and Socket.io munging code is at

🤦‍♂️ I searched the wrong flexdash repo.

Thanks @tve 

I will look at what you have posted with fresh eyes tomorrow.

**Comment by tve** [comment]
I'm happy to help, but the stack trace in the original post doesn't mean much to me.

**Comment by Steve-Mcl** [comment]
> This is related to the Express Session handling within Node-RED - which our authentication plugin drives in someway to cause this incompatibility.

> ### Steps To Reproduce
> * Install Flexdash within a FF hosted Node-RED
> * Add the example flow flexdash provides
> * Deploy and attempt to access the dashboard

Nick, you dont explicitly mention SSO (in the "steps to reproduce") but you do mention "authentication plugin" in the "Current Behavior" section.  Am I right in thinking when you captured this stack trace, it was on your local dev machine with SSO enabled?

Also, did you also see this issue with Basic Auth?

**Comment by Steve-Mcl** [comment]
> I'm happy to help, but the stack trace in the original post doesn't mean much to me.

Thanks Thorsten. We would love to have FlexDash running on our Cloud service & show it off to users so I appreciate that you are happy to help.

If we uncover something that pin points the issue and need your support, I will reach out.

Regards, Steve.

PS: If you wish to see this issue in action, there is a free 1 month trial available on flowforge.com (or get in touch and I can demonstrate it).

**Comment by knolleary** [comment]
I have identified the issue. This is not specific to FlowForge - any Node-RED with `adminAuth` configured to use an OAuth strategy will hit it.

When `adminAuth` is configured for oauth, it sets `req.session` to an instance of `express-session`. This includes a hook once the request is completed to trigger a refresh of the cookie (by calling `req.session.touch()`).

Flexdash uses `cookie-session` (https://github.com/flexdash/node-red-flexdash/blob/main/flexdash-dashboard.js#L278)

This sets `req.session` to a new session object with every request that hits the flexdash express app. In doing so it overwrites the existing `req.session` value with a different session implementation that doesn't include `touch`. So when `express-session` runs its code to refresh the cookie, `req.session` is no longer the one it expected.

There are a number of long-open issues around supporting multiple session objects in express to help support this type of application co-existence - unfortunately they have not got anywhere with proposed PRs unmerged.

The workaround will be a code change in flexdash to add some conditional logic around where it sets up `req.session` to only do so if no such object already exists. I have verified that works locally, but not done an exhaustive test of flexdash's cookie handling to ensure it all hangs together as expected. I will open a PR on flexdash shortly.
